File: problems/problems_438/Solution.cpp
//go:build ignore
#include "cpp/common/Solution.h"


using namespace std;
using json = nlohmann::json;

class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        vector<int> ans;
        vector<int> counts(26, 0);
        for (auto c: p) {
            counts[c - 'a']--;
        }
        int diff = 0;
        for (auto c: counts) {
            diff += c != 0;
        }
        int m = static_cast<int>(s.size()), n = static_cast<int>(p.size());
        auto helper = [&counts](int key, int val)-> int {
            bool before = counts[key] == 0;
            counts[key] += val;
            if (before) {
                return 1;
            }
            if (counts[key] == 0) {
                return -1;
            }
            return 0;
        };
        for (int i = 0; i < m; i++) {
            diff += helper(s[i] - 'a', 1);
            if (i >= n - 1) {
                if (diff == 0) {
                    ans.push_back(i - n + 1);
                }
                diff += helper(s[i - n + 1] - 'a', -1);
            }
        }
        return ans;
    }
};

json leetcode::qubh::Solve(string input_json_values) {
	vector<string> inputArray;
	size_t pos = input_json_values.find('\n');
	while (pos != string::npos) {
		inputArray.push_back(input_json_values.substr(0, pos));
		input_json_values = input_json_values.substr(pos + 1);
		pos = input_json_values.find('\n');
	}
	inputArray.push_back(input_json_values);

	Solution solution;
	string s = json::parse(inputArray.at(0));
	string p = json::parse(inputArray.at(1));
	return solution.findAnagrams(s, p);
}

File: problems/problems_438/solution.rs
use serde_json::{json, Value};

pub struct Solution;

impl Solution {
    pub fn find_anagrams(s: String, p: String) -> Vec<i32> {
        let mut ans: Vec<i32> = Vec::new();
        let m = s.len();
        let n = p.len();
        let mut counts = vec![0; 26];
        for c in p.chars() {
            counts[c as usize - 'a' as usize] -= 1;
        }
        let mut diff = 0;
        for &v in counts.iter() {
            if v != 0 {
                diff += 1;
            }
        }
        let mut helper = |key: usize, val: i32| -> i32 {
            let before = counts[key] == 0;
            counts[key] += val;
            if before {
                1
            } else if counts[key] == 0 {
                -1
            } else {
                0
            }
        };
		let chars: Vec<char> = s.chars().collect();
        for i in 0..m {
            diff += helper(chars[i] as usize - 'a' as usize, 1);
            if i >= n - 1 {
                if diff == 0 {
                    ans.push(i as i32 - n as i32 + 1);
                }
                if let Some(index) = (i as isize).checked_sub(n as isize - 1) {
                    diff += helper(chars[index as usize] as usize - 'a' as usize, -1);
                }
            }
        }
        ans
    }
}

#[cfg(feature = "solution_438")]
pub fn solve(input_string: String) -> Value {
    let input_values: Vec<String> = input_string.split('\n').map(|x| x.to_string()).collect();
    let s: String = serde_json::from_str(&input_values[0]).expect("Failed to parse input");
    let p: String = serde_json::from_str(&input_values[1]).expect("Failed to parse input");
    json!(Solution::find_anagrams(s, p))
}
